### ISSUE TITLE
Separate drafts into own repository

### ISSUE THREAD

**Issue by Quintus** [Secretchronicles/TSC#161]
I think the SVG stuff in the `drafts/` folder should be taken out of this repository and added to another one. It does not really belong to the code, but is referencial art for artists. Also it giantly blows up this repository’s size.

Vale,
Quintus

**Comment by Bugsbane** [comment]
Agreed on using svgz. We really do need to have an easy way of finding the svg file that made up a particular png though. Imagine if every time you wanted to make an edit to the code, and all of the files were organized into folders based on who wrote them, and most of them, usally all in one big folder, and most of them were missing. That's basically what we have now with art assets, and it heavily discourages going back and improving any pre-existing graphics.

Most of the time, I just end up recreating things from scratch as it's faster, however that doesn't encourage a graphical codebase that encourages new artists. For those brave enough to try, it's also harder to learn, as you can't usually examine the pre-existing work of others. For those who still submit new graphics, it means that they're less consistent with current graphics than they would otherwise be, as they can't copy anything (eg eyes, mouths, live path effect styles, etc).

**Comment by datahead8888** [comment]
We really should look at organizing the svg's based on content type and not author.

Perhaps we could have a separate folder that contains a hierarchy of links pointing to these content-organized-svg's.  This would allow people to browse by author if they wished.  It would have to be maintained, however.

**Comment by Bugsbane** [comment]
Most of the time what I really need, is to be able to find the SVG that was used to make a specific PNG pixmap. What would speed that up would be SVG's with the same file name (barring file extension) in an identical folder structure, rather than just throwing all the SVG's together in some other folder structure that has logic, but doesn't match the use case .

For example, the SVG that made /pixmaps/ground/green_1/hedges/big_2.png would be found somewhere like /drafts/pixmaps/ground/green_1/hedges/big_2.svg or in a repository specifically for source material in /pixmaps/ground/green_1/hedges/big_2.svg.

What takes so much time is trying to match up which folder the corresponding svg exists in (if it does at all).

**Comment by Quintus** [comment]
Bugsbane <notifications@github.com> writes:

> For example, the SVG that made
> /pixmaps/ground/green_1/hedges/big_2.png would be found somewhere like
> /drafts/pixmaps/ground/green_1/hedges/big_2.svg or in a repository

I fully agree with this. We should just reorganise the drafts in such a
way that the folder structure mirrors the folder structure under
data/pixmaps. That way it’ſ easy to find the target SVG if available. Still,
the drafts should go into a separate repository to reduce the size of
the main repository, especially when we start adding music/sound source
files as well.

In order to keep the author information, I suggest to use SVG’s metadata
facilities. In Inkscape, it’ſ fairly easy to add information on the
author, creation date, and even license directly in the SVG, though
regarding license information I don’t think it is enough as anybody can
easily edit the metadata — separate license confirmation thus will
continue to be required in lieu of a cryptographic signature. However,
author information is something that should really be carried along for
our own informational purposes.

Valete,
Quintus

**Comment by datahead8888** [comment]
Yes, maintaining a separate folder structure based on authors with links to the main entries organized by content type would probably be too tedious to maintain.  People can use search tools if they want to find an author's works.

**Comment by Quintus** [comment]
> very rigorous about checking future commits to make sure this is maintained in the future

I guess it’s pretty easy to write a script that validates the presence of the metadata. I can code one up if this is desired.

Vale,
Quintus

**Comment by kirbyfan64** [comment]
@Quintus Well, if you can't, I can try, but I need to know what language to write it in!

**Comment by Quintus** [comment]
As we use Ruby for most of our scripting needs, this would probably be the language of choice. Use the REXML library from Ruby’s standard library and target the metadata nodes with an XPath to check if they’re there. One could even extract the text and see if it is a CC license...

Vale,
Quintus

**Comment by datahead8888** [comment]
@Quintus - are you thinking this new script would be part of the build system in order to give warnings (ie invoked from the CMake system)?

**Comment by Quintus** [comment]
datahead8888 <notifications@github.com> writes:

> @Quintus - are you thinking this new script would be part of the build
> system in order to give warnings (ie invoked from the CMake system)?

I’d just define a task in the Rakefile so that you can at any time just
issue

  $ rake checksvgs

and it would print the names of those files that do not adhere to the
standards. Making it part of the buildsystem is overkill in my opinion.

Vale,
Quintus

**Comment by Quintus** [comment]
With regard to the original post, I suggest the following procedure:

1. Wait for the SFML port to be finished and merged into `devel`. I don’t like redoing large parts of the repository and then merging in a really large branch. I have a certain respect for the number of conflicts this may cause.
2. Clone the current TSC repository into a new repository named “TSC-graphics” or so.
3. Delete the `drafts/` folder from the TSC repository using a regular `git rm` command. That way we don’t lose history.
4. From the TSC-graphics repository, delete — exactly oppositely — the entire `tsc/` and `mruby` folders.
5. In the TSC repository, move everything inside the `tsc/` folder one level up. In the TSC-graphics repository, move everything inside the `drafts/` folder one level up.

At this stage, we have separated the two repositories apart, and both share the same history, but their current set of files is reduced. However, if you `git clone` the TSC repository, you’ll still get all the graphics files in the Git history. I suggest that you prevent that manually for the moment then by using Git’s `--depth` option on cloning with a sane value like 10. The more the game advances, the more commits are made, and when we are some hundred commits away from the splitting point of the two repositories, we can decide to cut off the common part before the splitting point from the TSC repository (*not* from the TSC-graphics repository, we don’t want to lose it completely). 

After cutting off the common part of the history from the TSC repository, then the graphics sources are really totally gone from that repository. On the opposite, we’d lose information for `git bisect`, which is why I don’t want to do this immediately after splitting, but rather some more distant time in the future so that `git bisect` remains useful to us (this is based on the assumption that it is unlikely to find really old commits in a bisecting session).

Immediately after splitting the repositories we can reorder the structure of the new TSC-graphics repository so it matches the structure inside the game’s `data/pixmaps` directory.

Opions on this?

Valete,
Quintus

**Comment by Bugsbane** [comment]
> so it matches the structure inside the game’s data/pixmaps directory

Music and sfx too, please, not just pixmaps. ;) Maybe call it TSC-media-assets?

**Comment by Quintus** [comment]
@datahead8888 suggested in IRC to use the more brutal hammer and rewrite Git history directly as follows:

1. Clone the repository three times
2. In the first clone, remove the draft files recursively from the Git history
3. In the second clone, remove the code files recursively from the Git history
4. Forcepush the first clone over the current TSC repository
5. Use the second clone to create the new TSC-drafts repository
6. Use the third clone as a backup TSC-old repository

This approach maintains bisecting ability, and it allows us to get rid of the large `drafts/` directory right now (after the SFML port has been completed, see above). The repositories will then look as if they have ever been two different ones, and we won’t lose history of any single file (the history is just “split up” between the two repositories then).

I am not yet sure whether this is better, and removing such a large amount of files from Git history may be complicated. I need to do a local test of that.

> Music and sfx too, please, not just pixmaps. ;) Maybe call it TSC-media-assets?

Yes, of course. But note that at the current point of time, we don’t have source files for music and sound effects like we have for the pixmaps. That is, we have SVGs, but we do not have Ardour/someotherprogram source files or at least FLAC files.

In any case, the PNG files remain in the main repository with both approaches outlined.

Vale,
Quintus
